### PR TITLE
[bitnami/wordpress] Support for cronjobs in wordpress

### PR DESCRIPTION
**Pull request by lotyp**
### Description of the change

This pull request introduces CronJob support to the Bitnami WordPress Helm chart. It adds a new section in the values.yaml file for configuring CronJobs

### Benefits

The integration of CronJob support provides several benefits:

* Enables automation of routine WordPress tasks, enhancing the functionality of the WordPress deployment.
* Offers flexibility to chart users to define their own custom CronJobs according to their specific needs.
* Ensures that CronJobs are consistently and correctly configured, following the Bitnami chart standards.

### Possible drawbacks

* Introducing CronJobs adds an additional layer of complexity to the chart, which might require users to have a bit more understanding of Kubernetes CronJobs.
* In cases where CronJobs are not configured properly, it might lead to unintended behaviors or performance issues.

### Applicable issues

* Initial implementation, no existing issues related to this feature.

### Additional information

* This change has been tested to ensure that it does not interfere with the existing functionalities of the WordPress chart and follows the Bitnami best practices for Helm chart development.

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### Linked issues:

https://github.com/bitnami/charts/issues/21246

### CODE REVIEW

**Comment by lotyp** [comment]
I have tested this functionality by manually disabling default cron behavior in `wp-config.php`

**Review by jevgenijsblaus — APPROVED** [review]

File: bitnami/wordpress/values.yaml [hunk 1]
```diff
@@ -1217,3 +1217,93 @@ externalCache:
...
+    ## @param cronJobs.scheduler.schedule Schedule for the WordPress scheduler cron job
+    ##
+    schedule: "*/2 * * * *"
+    ## @param cronJobs.scheduler.command Override default container command (useful when using custom images)
+    ##
+    command: []
+    ## @param cronJobs.scheduler.args Override default container args (useful when using custom images)
+    ##
+    args: []
+    ## @param
```
Comment (juan131) [inline bitnami/wordpress/values.yaml:1240]:
```suggestion
```

File: bitnami/wordpress/values.yaml [hunk 2]
```diff
@@ -1217,3 +1217,93 @@ externalCache:
...
+        digest: ""
+        pullPolicy: IfNotPresent
+        ## Optionally specify an array of imagePullSecrets.
+        ## Secrets must be manually created in the namespace.
+        ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
+        ## e.g:
+        ## pullSecrets:
+        ##   - myRegistryKeySecretName
+        ##
+        pullSecrets: []
```
Comment (juan131) [inline bitnami/wordpress/values.yaml:1309]: It is not allowed to include non-Bitnami images on Bitnami charts. As an alternative you may use `bitnami/os-shell` image as other charts do.

File: bitnami/wordpress/templates/cronjobs.yaml [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            - name: wordpress-data
+              {{- if .Values.persistence.enabled }}
+              persistentVolumeClaim:
+                claimName: {{ .Values.persistence.existingClaim | default (include "common.names.fullname" .) }}
+              {{- else }}
+              emptyDir: {}
+              {{- end }}
+            {{- if .Values.extraVolumes }}
+            {{- include "common.tplvalues.render" (dict "value" .Values.extraVolumes "context" $) | nindent 8 }}
+            {{- end }}
```
Comment (juan131) [inline bitnami/wordpress/templates/cronjobs.yaml:249]: Why do we need all these env. vars, mounted configuration, etc.?

File: bitnami/wordpress/values.yaml [hunk 4]
```diff
@@ -1217,3 +1217,93 @@ externalCache:
...
   ##
   port: 11211
+
+## @section CronJob parameters
+##
+
+## Configure cron jobs used to perform periodic tasks
+##
+cronJobs:
+  scheduler:
```
Comment (juan131) [inline bitnami/wordpress/values.yaml:1227]: It seems there's only one cronjob you're actually defining. Do you expect to add support for more? Isn't this sth that can be already covered with `extraDeploy` parameter?

File: bitnami/wordpress/templates/cronjobs.yaml [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  failedJobsHistoryLimit: {{ .Values.cronJobs.scheduler.failedJobsHistoryLimit }}
+  successfulJobsHistoryLimit: {{ .Values.cronJobs.scheduler.successfulJobsHistoryLimit }}
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+            {{- if .Values.cronJobs.scheduler.podLabels }}
+              {{- include "common.tplvalues.render" ( dict "value" .Values.cronJobs.scheduler.podLabels "context" $ ) | nindent 12 }}
+            {{- end }}
```
Comment (juan131) [inline bitnami/wordpress/templates/cronjobs.yaml:28]: Use "common.labels.standard" helper merging `.Values.commonLabels` with `.Values.cronJobs.scheduler.podLabels`. Also, consider annotations.
See how others charts do it:

- https://github.com/bitnami/charts/blob/main/bitnami/redmine/templates/cronjob.yaml#L24-L30

File: bitnami/wordpress/templates/cronjobs.yaml [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+metadata:
+  name: {{ include "common.names.fullname" . }}-scheduler
+  namespace: {{ .Release.Namespace | quote }}
+  labels: {{- include "common.labels.standard" ( dict "customLabels" .Values.commonLabels "context" $ ) | nindent 4 }}
+    {{- if .Values.commonAnnotations }}
+  annotations: {{- include "common.tplvalues.render" ( dict "value" .Values.commonAnnotations "context" $ ) | nindent 4 }}
+    {{- end }}
+spec:
+  schedule: {{ .Values.cronJobs.scheduler.schedule | quote }}
+  concurrencyPolicy: {{ .Values.cronJobs.scheduler.concurrencyPolicy }}
```
Comment (juan131) [inline bitnami/wordpress/templates/cronjobs.yaml:18]: Do we want to allow concurrency?

File: bitnami/wordpress/templates/cronjobs.yaml [hunk 7]
```diff
@@ -0,0 +1,250 @@
+{{- /*
+Copyright VMware, Inc.
+SPDX-License-Identifier: APACHE-2.0
+*/}}
+
+{{- if .Values.cronJobs.scheduler.enabled }}
+apiVersion: batch/v1
```
Comment (juan131) [inline bitnami/wordpress/templates/cronjobs.yaml:7]:
```suggestion
apiVersion: {{ include "common.capabilities.cronjob.apiVersion" . }}
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for this contribution! I added a few comments

**Comment by brandan-schmitz** [comment]
Any change this can be re-opened as it was never completed but is still a very useful thing to get added?

**Comment by carrodher** [comment]
Thank you for bringing this issue to our attention. We appreciate your involvement! If you're interested in contributing a solution, we welcome you to create a pull request. The Bitnami team is excited to review your submission and offer feedback. You can find the contributing guidelines [here](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md).

Your contribution will greatly benefit the community. Feel free to reach out if you have any questions or need assistance.

**Comment by lotyp** [comment]
> Any change this can be re-opened as it was never completed but is still a very useful thing to get added?

I remember about this PR 😅

Unfortunately right now I don't have any k8s cluster available to test this out, but if I will have time I could finish it.

**Pull request closed without merging**